File: LeetCode/0168-Excel Sheet Column Title/main.cpp
//
// Created by 薛智钧 on 2020/3/31.
//
#include <string>
#include <iostream>
using namespace std;

/*
 * Given a positive integer, return its corresponding column title as appear in an Excel sheet.

For example:

    1 -> A
    2 -> B
    3 -> C
    ...
    26 -> Z
    27 -> AA
    28 -> AB
   701 ->ZY
 * */

class Solution {
public:
    // 从低位往高位求，每进一位，则把原数缩小26倍，再对26取余，之后减去余数
    // 对于小于26的数字，只需要对26取余，然后减去1，加上字符A即可
    string convertToTitle(int n) {
        if (n < 1) return "";
        string res;
        while (n){
            int mod = (n - 1) % 26; // 因为要与'A'相加，因此要先－1再取余
            res += (char)(mod + 'A'); // 'Z' = 'A' + 25
            n -= (mod + 1); // 减去真正的余数
            n /= 26;
        }
        reverse(res.begin(), res.end());
        return res;
    }

    // 递归法
    string convertToTitle_R(int n){
        return n == 0 ? "" :
        convertToTitle_R((n - 1) / 26) + (char)((n - 1) % 26 + 'A');
    }
};

int main(){
    int n = 99;
    Solution sol;
    cout << sol.convertToTitle(n);
}

File: LeetCode/0168-Excel Sheet Column Title/main.py
"""  
Given a positive integer, return its corresponding column title as appear in an Excel sheet.

For example:

    1 -> A
    2 -> B
    3 -> C
    ...
    26 -> Z
    27 -> AA
    28 -> AB 
    ...
Example 1:

Input: 1
Output: "A"
"""

"""  
Now we can see that ABCD＝A×26³＋B×26²＋C×26¹＋D＝1×26³＋2×26²＋3×26¹＋4

But how to get the column title from the number? We can't simply use the n%26 method because:

ZZZZ＝Z×26³＋Z×26²＋Z×26¹＋Z＝26×26³＋26×26²＋26×26¹＋26

We can use (n-1) % 26 instead, then we get a number range from 0 to 25.
"""

class Solution:
    def convertToTitle(self, n: int) -> str:
        res = []
        while n > 0:
            res.append(chr((n - 1) % 26 + ord('A')))
            n = (n - 1) // 26
            
        res.reverse()
        return ''.join(res)